token/js/src/extensions/tokenMetadata/actions.ts

```diff
@@ -0,0 +1,387 @@
+import type { ConfirmOptions, Connection, PublicKey, Signer, TransactionSignature } from '@solana/web3.js';
+import type { Field, TokenMetadata } from '@solana/spl-token-metadata';
+import {
+    createEmitInstruction,
+    createInitializeInstruction,
+    createRemoveKeyInstruction,
+    createUpdateAuthorityInstruction,
+    createUpdateFieldInstruction,
+    pack,
+    unpack,
+} from '@solana/spl-token-metadata';
+import { sendAndConfirmTransaction, SystemProgram, Transaction } from '@solana/web3.js';
+import { TokenAccountNotFoundError } from '../../errors.js';
+
+import { TOKEN_2022_PROGRAM_ID } from '../../constants.js';
+import { getSigners } from '../../actions/internal.js';
+import { unpackMint } from '../../state/mint.js';
+import { getExtensionData, ExtensionType } from '../extensionType.js';
+import { updateTokenMetadata } from './state.js';
+
+/**
+ * Calculates additional lamports need for variable length extension
+ *
+ * @param connection       Connection to use
+ * @param address          Mint Account
+ * @param getNewMetadata   Function to get new metadata, given current
+ * @param programId        SPL Token program account
+ *
+ * @return lamports to send
+ */
+async function getAdditionalRentForNewMetadata(
+    connection: Connection,
+    address: PublicKey,
+    getNewMetadata: (x: TokenMetadata) => TokenMetadata,
+    programId = TOKEN_2022_PROGRAM_ID
+): Promise<number> {
+    const info = await connection.getAccountInfo(address);
+    const mint = unpackMint(address, info, programId);
+
+    if (!info) {
+        // unpack mint does error checking on info internally
+        // Check here for typescript and just in-case
+        throw new TokenAccountNotFoundError();
+    }
+
+    const data = getExtensionData(ExtensionType.TokenMetadata, mint.tlvData);
+
+    const currentDataLen = data ? data.length : 0;
+
+    const newMeta = data === null ? getNewMetadata({} as TokenMetadata) : getNewMetadata(unpack(data));
+    const newDataLen = pack(newMeta).length;
+
+    let newAccountLen = info.data.length + (newDataLen - currentDataLen);
```

Comment:
We might need to roll something like this in the JS library, for general use:

https://github.com/solana-labs/solana-program-library/blob/b49a7ceebb900f64ec0389b0ce1e35579a5c2ade/token/program-2022/src/extension/mod.rs#L371-L391

Reply:
I've put something together - let me know if its the right approach. 

Some things i want to check:

1. You mention a 4 byte discriminator - but I have a 2 byte one. The test cases passing leads me to believe its correct - but should it be 4?

2. Do we need to recalculate the base length in the js implementation?  I believe just taking the account length and subtracting the current extension length and adding the length length should be sufficient?

I started calling [`getExtensionTypes`](https://github.com/solana-labs/solana-program-library/blob/master/token/js/src/extensions/extensionType.ts#L207) and feeding into [`getLen`](https://github.com/solana-labs/solana-program-library/blob/master/token/js/src/extensions/extensionType.ts#L166C10-L166C16) - but needs some extra logic around variable length extensions - which i'm unsure we need.

Reply:
> 1. You mention a 4 byte discriminator - but I have a 2 byte one. The test cases passing leads me to believe its correct - but should it be 4?

Oops, my bad. 2 is correct.
https://github.com/solana-labs/solana-program-library/blob/5da184aed56cda46b0006ebf3ea5d617ef0501a8/token/program-2022/src/extension/mod.rs#L902

> 2. Do we need to recalculate the base length in the js implementation?  I believe just taking the account length and subtracting the current extension length and adding the length length should be sufficient (which is what I did)?

The base length is static for mints or accounts. What the function needs to do is check if the extension exists, if it does calculate the size difference, if it doesn't get the new size.